src/x2attr.c

```diff
@@ -0,0 +1,649 @@
+    \ingroup libgerbv
+*/
+
+#include "gerbv.h"
+
+#include <ctype.h>
+
+#include "common.h"
+#include "gerber.h"
+#include "gerb_stats.h"
+
+#include "x2attr.h"
+
+
+//#define dprintf if(DEBUG) printf
+#define dprintf if(1) printf
+
+
+/* Dynamic attribute dictionary state while parsing RS274-X2. 
+
+Implementation note:
+
+Attribute names and values are interned using g_intern_string(), which allows hanshing and equality
+testing to be performed on the *pointers* rather than possibly lengthy strings.
+
+Three of these hash tables are referenced from gerb_state_t, one for each type.  When inserting a
+new name, the *other* two tables are checked to ensure uniqueness over all attribute types, as
+required by the standard.
+
+
+*/
+typedef struct x2attr_dict {
+    x2attr_type_t type;
+    GHashTable * hashtable;
+} x2attr_dict_t;
+
+/* Attribute table (state) retained after parsing, which
+represents a snapshot of attributes associated with image, aperture
+or net objects, from the above dictionary object(s).
+
+Library users access using x2attr_get_image_attr() etc., modify using
+x2attr_set_image_attr() etc.  This is also used when importing PnP and
+IPC-D-356 files.
+
+When parsing RS274-X2, a private implementation is used to create these
+according to the standard.
+
+This is stored as a (possibly) sorted table of tuples (const char * key, const char * value).
+NOTE: sorting is by the actual
+key value, not the pointer which is essentially random.  The purpose of sorting is
+to make search more efficient if necessary, but primarily to create a consistent
+canonical attribute output order when writing RS274-X2.  This helps with testing.
+
+We could store a hash table, but that's a fairly heavyweight structure,
+so use this vector-based approach instead.  The initial link pointer
+indicates the next table to search if this one comes up empty.  That's used to
+point to the aperture attributes if this is a net attribute table.
+
+This is a variable length struct, so use g_malloc.  Most of the time, its
+size is known at creation time.  Otherwise, g_realloc is used to resize.
+We don't bother to track allocation - it's as big as it needs to be.
+
+Expect not that many different attribute keys, but a lot of objects and
+values, so a small slow struct is better than a big fast one.
+
+*/
+typedef struct x2attr_state {
+    gboolean sorted;        // TRUE when keys are sorted in canonical order.  This is
+                            // only set when _sort() called.  Reset on any modification.
+    unsigned size;          // Number of elements in following
+    struct x2attr_state * next; // Aperture attributes if this is for net (object), etc.
+    struct {
+        const char * key;   // Pointer to interned UTF-8 null terminated string.
+                            // Sort by key if 'sorted'.
+        const char * value; // Pointer to interned UTF-8 null terminated string
+    } t[1]; // 1st of 'size' elements.
```

Comment:
Why is it called `t`?  Could we be more descriptive with the name?  Like `attributes` or something?